src/url_aggregator.cpp

```diff
@@ -26,6 +29,63 @@ inline void url_aggregator::set_scheme(std::string_view new_scheme) noexcept {
   // TODO: implement
 }
 
+bool url_aggregator::set_protocol(const std::string_view input) {
+  (void) input;
+  // TODO: Implement
+  return false;
+}
+
+bool url_aggregator::set_username(const std::string_view input) {
+  if (cannot_have_credentials_or_port()) { return false; }
+  size_t username_start = components.protocol_end + 3;
+  size_t username_length = components.username_end - username_start;
+  buffer.erase(username_start, components.username_end);
+
+  std::string encoded_input = ada::unicode::percent_encode(input, character_sets::USERINFO_PERCENT_ENCODE);
```

Comment:
Here we could avoid the temporary string creation, but the set_username function is probably not performance critical, so it is fine. I would add a remark in the comment that this can probably be further optimized. Maybe. (You don't have to do it.)